_drafts/device-logging.md

```diff
@@ -0,0 +1,295 @@
+
+I've been lucky in that embedded Android devices are generally not very disk space constrained, so I’ve had the benefit of having 10's of MBs of logs to store (and comb) through. But even so, that would only constitute maybe 2-3 hours of logs, depending on how busy the device was, and there was a constant trade-off with how much information was being logged.
+
+### On-location v. Remote Collection
+
+One of my favorite often overlooked questions when you develop a new embedded device product is, "how do I get the logs off of my production device and onto my computer"? During development, this could be as easy as connecting a cable directly to the device and downloading the logs files off the filesystem directly. But when the device is in the field, it’s not so simple.
+
+I distinctly remember going to out to a customer’s store and connecting to their network so I could pull logs off of a network connected device, just to diagnose an issue they were having. This is of course, unsustainable if you have hardware that’s distributed across multiple locations and want to collect logs more frequently than each time you have physical access to the device.
+
+In that case, you’ll want to build a way to upload logs to a backend somewhere, account for all the storage and privacy-related concerns and regulations, and then a way to download those logs from your computer to analyze them.
+
+### Triggering Collection
+
+Naively, the easiest way to upload logs is have the user do it. If the device has a UI, it’s as simple as building a button for the user to click when they’re having problems &mdash; think about the "Shake to Report Feedback" or "Submit Feedback" feature on many apps. This way, you can be upfront about the data you’re collecting right, at the time of collection.
+
+Sadly, this depends on the user though, and users are notoriously unreliable. If they’re a technical user, they might not even need an UI! — they might be happy using a CLI tool instead. But if they’re non-technical, even asking them to navigate to a certain page might be a huge burden.
+
+The first alternative that comes to mind is a push system for retrieving logs &mdash; send a notification from the server to the device, and the device will upload the logs when it receives the message. This can take the form of an actual push notification mechanism with a connected web socket, or something as simple as having the device perform a poll every once in a while and collecting the log if the poll tells it to.
```

Comment:
This section could use a bit of a re-write imo. 
> The first alternative that comes to mind is a push system for retrieving logs

but then you immediately talk about a polling solution. 

Personally I'd reverse this and say something along the lines of:
The easiest solution is to have your device poll periodically to see if it needs to upload logs. Alternatively, a more advanced system would be to use a push notification system utilizing a websocket or MQTT.

Reply:
Well a polling solution is basically a push in this context - the developer has to push an request to the device. The mechanism for retrieving that request is either a polling solution or a pushing solution. 

But probably mixing up poll and push in this way won't be as obvious.

Reply:
gotcha, yeah I see what you mean now. I wasn't thinking about it in that way.